peer/peer.go

```diff
@@ -84,43 +100,204 @@ func New(ip string, port int) (*Peer, error) {
+// opens a new stream with this peer.
+// This should be passed as the second argument to SetStreamHandler() after this
+// peer is initialized.
 func (p *Peer) Receive(s net.Stream) {
-	log.Debug("Setting basic stream handler.")
-	p.doCumulus(s)
-}
+	// Get remote peer's full multiaddress
+	remoteMA, err := NewMultiaddr(
+		s.Conn().RemoteMultiaddr(), s.Conn().RemotePeer())
+	if err != nil {
+		log.WithError(err).Error(
+			"Failed to obtain valid remote peer multiaddress")
+		return
+	}
 
-// Communicate with peers.
-// TODO: Update this to do something useful. For now it just reads from the
-// stream and writes back what it read.
-func (p *Peer) doCumulus(s net.Stream) {
-	buf := bufio.NewReader(s)
-	str, err := buf.ReadString('\n')
+	// Add the remote peer to this peer's subnet
+	err = p.subnet.AddPeer(remoteMA, s)
 	if err != nil {
-		log.Error(err)
+		// Subnet is full, advertise other available peers and then close
+		// the stream
+		log.WithError(err).Debug("Peer subnet full. Advertising peers...")
+		msg := message.NewPushMessage(message.ResourcePeerInfo, p.subnet.StringMultiaddrs())
+		msgErr := msg.Write(s)
+		if msgErr != nil {
+			log.WithError(err).Error("Failed to send ResourcePeerInfo")
+		}
+		s.Close()
 		return
 	}
+	go p.Listen(remoteMA, s)
+}
+
+// Connect adds the given multiaddress to p's Peerstore and opens a stream
+// with the peer at that multiaddress if the multiaddress is valid, otherwise
+// returns error. On success the stream and corresponding multiaddress are
+// added to this peer's subnet.
+func (p *Peer) Connect(peerma string) (net.Stream, error) {
+	peerid, targetAddr, err := extractPeerInfo(peerma)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the peer's address in this host's PeerStore
+	p.Peerstore().AddAddr(peerid, targetAddr, pstore.PermanentAddrTTL)
+
+	// Open a stream with the peer
+	stream, err := p.NewStream(context.Background(), peerid, CumulusProtocol)
+	if err != nil {
+		return nil, err
+	}
+	err = stream.SetDeadline(time.Now().Add(Timeout))
+	if err != nil {
+		log.WithError(err).Error("Failed to set read deadline on stream")
+	}
+
+	mAddr, err := ma.NewMultiaddr(peerma)
+	if err != nil {
+		stream.Close()
+		return nil, err
+	}
+
+	err = p.subnet.AddPeer(mAddr, stream)
+	if err != nil {
+		stream.Close()
+	}
+
+	go p.Listen(mAddr, stream)
+	return stream, nil
+}
+
+// Broadcast sends message to all peers this peer is currently connected to
+func (p *Peer) Broadcast(m message.Message) error {
+	return errors.New("Function not implemented")
+}
+
+// Request sends a request to a remote peer over the given stream.
+// Returns response if a response was received, otherwise returns error.
+// You should typically run this function in a goroutine
+func (p *Peer) Request(req message.Request, s net.Stream) (*message.Response, error) {
+	// Set up a listen channel to listen for the response (remove when done)
+	lchan := p.newListener(req.ID)
```

Comment:
I don't like this `newListener` business. I think we should partition better based on stream